src/sspi/pku2u/cert_utils/win_extraction.rs

```diff
@@ -0,0 +1,329 @@
+
+    let blob_type_wide = string_to_utf16(BCRYPT_RSAFULLPRIVATE_BLOB);
+
+    // The first call need to determine the size of the needed buffer for the private key
+    // https://learn.microsoft.com/en-us/windows/win32/api/ncrypt/nf-ncrypt-ncryptexportkey
+    // If pbOutput parameter is NULL, this function will place the required size in the pcbResult parameter.
+    let status = NCryptExportKey(
+        private_key_handle as _,
+        0,
+        blob_type_wide.as_ptr() as *const _,
+        null(),
+        null::<u8>() as *mut _,
+        0,
+        &mut private_key_buffer_len,
+        0,
+    );
+
+    if status != 0 {
+        NCryptFreeObject(private_key_handle);
+
+        return Err(Error::new(
+            ErrorKind::InternalError,
+            format!(
+                "Cannot extract certificate private key: unsuccessful extraction: {:x?}",
+                status
+            ),
+        ));
+    }
+
+    let mut private_key_blob = vec![0; private_key_buffer_len as usize];
+
+    let status = NCryptExportKey(
+        private_key_handle as _,
+        0,
+        blob_type_wide.as_ptr() as *const _,
+        null(),
+        private_key_blob.as_mut_ptr(),
+        private_key_blob.len() as _,
+        &mut private_key_buffer_len,
+        0,
+    );
+
+    NCryptFreeObject(private_key_handle);
+
+    if status != 0 {
+        return Err(Error::new(
+            ErrorKind::InternalError,
+            format!(
+                "Cannot extract certificate private key: unsuccessful extraction: {:x?}",
+                status
+            ),
+        ));
+    }
+
+    let private_key = decode_private_key(&private_key_blob[0..private_key_buffer_len as usize])?;
+
+    Ok(private_key)
+}
+
+unsafe fn extract_client_p2p_certificate(cert_store: *mut c_void) -> Result<(Certificate, RsaPrivateKey)> {
+    let mut certificate = CertEnumCertificatesInStore(cert_store, null_mut());
+
+    while !certificate.is_null() {
+        let cert_der = from_raw_parts((*certificate).pbCertEncoded, (*certificate).cbCertEncoded as usize);
+        let cert: Certificate = picky_asn1_der::from_bytes(cert_der)?;
+
+        if !validate_client_p2p_certificate(&cert) {
+            let next_certificate = CertEnumCertificatesInStore(cert_store, certificate);
+
+            certificate = next_certificate;
+
+            continue;
+        }
+
+        let private_key = export_certificate_private_key(certificate);
+
+        CertFreeCertificateContext(certificate);
+
+        return Ok((cert, private_key?));
+    }
+
+    Err(Error::new(
+        ErrorKind::InternalError,
+        "Cannot find appropriate device certificate".into(),
+    ))
+}
+
+pub fn extract_client_p2p_cert_and_key() -> Result<(Certificate, RsaPrivateKey)> {
+    unsafe {
+        // "My\0" encoded as a wide string.
+        // More info: https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-certopenstore#remarks
+        let my: [u16; 3] = [77, 121, 0];
```

Comment:
I believe there are other system store names besides `My` such as `Root`, `TrustedPeople`, `TrustedPublisher`…
The link is not explaining the rationale behind choosing specifically `My` here. Are we sure we never need to look into other stores?

Reply:
I've noticed that a `P2P` certificate is always a personal one. I had the same in the further dev-testing. I haven't found information that P2P certificates are always there. According to what I've had, I believe it's always there. 
 Through the `My` certificate store we can access personal certificates. So It's why it's in use

Reply:
I understand. Could you just add this information in a comment above? It’s really just to let a future reader know that the question has been raised and it’s not arbitrary :slightly_smiling_face:

Reply:
Done